src/main/java/org/jfrog/hudson/pipeline/common/Utils.java

```diff
@@ -392,25 +390,42 @@ public static String replaceTildeWithUserHome(String path) {
         return path.replaceFirst("^~", System.getProperty("user.home"));
     }
 
-    public static String getNpmExe(TaskListener listener, EnvVars env, Launcher launcher, String nodeTool) throws IOException, InterruptedException {
+    public static String getNpmExe(FilePath ws, TaskListener listener, EnvVars env, Launcher launcher, String nodeTool) throws IOException, InterruptedException {
+        Log logger = new JenkinsBuildInfoLog(listener);
+        String npmPath = "";
         // npm from tool
         if (StringUtils.isNotEmpty(nodeTool)) {
-            NodeJSInstallation nodeInstallation = getNpmInstallation(nodeTool);
-            if (nodeInstallation == null) {
-                listener.error("Couldn't find NodeJS tool '" + nodeTool + "'");
-                throw new Run.RunnerAbortedException();
-            }
-            Node node = ActionableHelper.getNode(launcher);
-            nodeInstallation = nodeInstallation.forNode(node, listener).forEnvironment(env);
-            String nodeExe = nodeInstallation.getExecutable(launcher);
-            return Paths.get(nodeExe).resolveSibling("npm").toString();
+            npmPath = getNpmFromTool(ws, logger, listener, env, launcher, nodeTool);
         }
         // npm from environment
         if (env.containsKey("NODEJS_HOME")) {
-            return Paths.get(env.get("NODEJS_HOME"), "bin", "npm").toString();
+            npmPath = ws.child(env.get("NODEJS_HOME")).child("bin").child("npm").getRemote();
+        }
+        logger.debug("Using npm executable from " + StringUtils.defaultIfEmpty(npmPath, "PATH"));
+        // If npmPath is empty, try to use npm from PATH
+        return npmPath;
+    }
+
+    private static String getNpmFromTool(FilePath ws, Log logger, TaskListener listener, EnvVars env, Launcher launcher, String nodeTool) throws IOException, InterruptedException {
+        NodeJSInstallation nodeInstallation = getNpmInstallation(nodeTool);
+        if (nodeInstallation == null) {
+            logger.error("Couldn't find NodeJS tool '" + nodeTool + "'");
+            throw new Run.RunnerAbortedException();
+        }
+        Node node = ActionableHelper.getNode(launcher);
+        nodeInstallation = nodeInstallation.forNode(node, listener).forEnvironment(env);
+        String nodeJsHome = nodeInstallation.forNode(node, listener).forEnvironment(env).getHome();
+        if (StringUtils.isBlank(nodeJsHome)) {
+            logger.error("Couldn't find NodeJS home");
+            throw new Run.RunnerAbortedException();
+        }
+        FilePath nodePath = ws.child(nodeJsHome).child("bin").child("node");
+        if (!nodePath.exists()) {
+            logger.error("Couldn't find node executable in path " + nodePath.getRemote());
+            throw new Run.RunnerAbortedException();
         }
-        // npm from path
-        return "";
+        env.override(NodeJSConstants.ENVVAR_NODEJS_PATH, nodePath.getParent().getRemote()); // Prepend NODEJS_HOME/bin to PATH
```

Comment:
Let's move the above comment to be above the line it describes.
It is easier to read.